File: src/test/java/kitchenpos/application/MenuServiceChangePriceTest.java
package kitchenpos.application;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.google.common.collect.ImmutableList;
import java.math.BigDecimal;
import java.util.UUID;
import kitchenpos.domain.Menu;
import kitchenpos.domain.MenuGroup;
import kitchenpos.domain.MenuGroupRepository;
import kitchenpos.domain.MenuRepository;
import kitchenpos.domain.Product;
import kitchenpos.domain.ProductRepository;
import kitchenpos.infra.PurgomalumClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class MenuServiceChangePriceTest extends AbstractApplicationServiceTest {

    private static final BigDecimal TEST_PRICE = BigDecimal.valueOf(1_000L);
    private static final String TEST_MENU_NAME = "dummyMenuName";
    private static final String TEST_PRODUCT_NAME = "dummyProductName";
    private static final String TEST_MENU_GROUP_NAME = "dummyMenuGroup";

    @Mock
    private PurgomalumClient mockClient;

    private MenuRepository menuRepository;
    private MenuGroupRepository menuGroupRepository;
    private ProductRepository productRepository;

    private MenuService service;

    @BeforeEach
    void setUp() {
        menuRepository = new MenuFakeRepository();
        menuGroupRepository = new MenuGroupFakeRepository();
        productRepository = new ProductFakeRepository();

        service = new MenuService(menuRepository, menuGroupRepository,
            productRepository, mockClient);
    }

    @DisplayName("메뉴 가격을 변경할 수 있다.")
    @Test
    void changePrice_success() {
        // given
        final Product product = productRepository.save(
            createProductRequest(TEST_PRODUCT_NAME, TEST_PRICE));
        final MenuGroup menuGroup = menuGroupRepository.save(
            createMenuGroupRequest(TEST_MENU_GROUP_NAME));
        final Menu menu = menuRepository.save(create(product, menuGroup));
        final Menu request = create(BigDecimal.valueOf(500L));

        // when
        final Menu actual = service.changePrice(menu.getId(), request);

        // then
        assertThat(actual.getPrice())
            .isEqualTo(request.getPrice());
    }

    private Menu create(final BigDecimal price) {
        return createMenuRequest(TEST_MENU_NAME, price, ImmutableList.of(),
            createMenuGroupRequest(TEST_MENU_GROUP_NAME), false);
    }

    @DisplayName("메뉴 가격을 변경할 때 가격이 없거나 음수면 예외를 발생시킨다.")
    @Test
    void changePrice_invalid_price_1() {
        // given
        final UUID menuId = UUID.randomUUID();
        final Menu negativePrice = create(BigDecimal.valueOf(-1));
        final Menu nullPrice = create(null);

        // when & then
        assertThatThrownBy(() -> service.changePrice(menuId, negativePrice))
            .isInstanceOf(IllegalArgumentException.class);

        assertThatThrownBy(() -> service.changePrice(menuId, nullPrice))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @DisplayName("메뉴의 가격을 변경할 때 가격이 menuProduct들의 수량과 product의 가격을 곱한 값의 총 합보다 크면 예외를 발생시킨다.")
    @Test
    void changePrice_invalid_price_2() {
        // given
        final Product product = productRepository.save(createProductRequest("product", TEST_PRICE));
        final MenuGroup menuGroup = menuGroupRepository.save(
            createMenuGroupRequest(TEST_MENU_GROUP_NAME));
        final Menu menu = menuRepository.save(create(product, menuGroup));
        final Menu request = create(BigDecimal.valueOf(10_000L));

        // when & then
        assertThatThrownBy(() -> service.changePrice(menu.getId(), request));

    }

    private Menu create(final Product product, final MenuGroup menuGroup) {
        return createMenuRequest(TEST_MENU_NAME, TEST_PRICE,
            ImmutableList.of(createMenuProductRequest(product, 1)), menuGroup, false);
    }
}

File: src/test/java/kitchenpos/application/MenuServiceChangePriceTest.kt
package kitchenpos.application

import io.kotest.assertions.throwables.shouldThrowAny
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeTypeOf
import io.mockk.every
import io.mockk.mockk
import java.math.BigDecimal
import kitchenpos.domain.Menu
import kitchenpos.domain.MenuGroupRepository
import kitchenpos.domain.MenuRepository
import kitchenpos.domain.ProductRepository
import kitchenpos.infra.PurgomalumClient
import kitchenpos.testsupport.FakeMenuGroupRepository
import kitchenpos.testsupport.FakeMenuRepository
import kitchenpos.testsupport.FakeProductRepository
import kitchenpos.testsupport.MenuFixtures
import kitchenpos.testsupport.ProductFixtures

class MenuServiceChangePriceTest : ShouldSpec({
    lateinit var menuRepository: MenuRepository
    lateinit var menuGroupRepository: MenuGroupRepository
    lateinit var productRepository: ProductRepository
    lateinit var purgomalumClient: PurgomalumClient
    lateinit var service: MenuService

    lateinit var savedMenu: Menu

    beforeTest {
        menuRepository = FakeMenuRepository()
        menuGroupRepository = FakeMenuGroupRepository()
        productRepository = FakeProductRepository()
        purgomalumClient = mockk {
            every { containsProfanity(any()) } returns false
        }


        savedMenu = menuRepository.save(
            MenuFixtures.createMenu(
                product = ProductFixtures.createProduct()
            )
        )

        service = MenuService(
            menuRepository,
            menuGroupRepository,
            productRepository,
            purgomalumClient,
        )
    }

    context("메뉴 수정") {
        should("성공") {
            // given
            val menuId = savedMenu.id
            val price = savedMenu.price - 1.toBigDecimal()
            val request = createRequest(
                price = price
            )

            // when
            val changedMenu = service.changePrice(
                menuId,
                request
            )

            // then
            changedMenu.id shouldBe menuId
            changedMenu.price shouldBe price
            changedMenu.isDisplayed shouldBe true
        }

        should("실패 - 메뉴 가격이 0원 미만인 경우") {
            // given
            val menuId = savedMenu.id
            val minusPrice = savedMenu.price * (-1).toBigDecimal()
            val request = createRequest(
                price = minusPrice
            )

            // when
            val exception = shouldThrowAny {
                service.changePrice(
                    menuId,
                    request
                )
            }

            // then
            exception.shouldBeTypeOf<IllegalArgumentException>()
        }

        should("실패 - 메뉴 가격이 메뉴 상품들의 가격 합보다 큰 경우") {
            // given
            val menuId = savedMenu.id
            val overPrice = savedMenu.price * (2).toBigDecimal()
            val request = createRequest(
                price = overPrice
            )

            // when
            val exception = shouldThrowAny {
                service.changePrice(
                    menuId,
                    request
                )
            }

            // then
            exception.shouldBeTypeOf<IllegalArgumentException>()
        }
    }
}) {
    companion object {
        private fun createRequest(
            price: BigDecimal = 1000.toBigDecimal()
        ): Menu {
            return Menu().apply {
                this.price = price
            }
        }
    }
}
